hazelcast/src/main/java/com/hazelcast/jet/JobInvocationObserver.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet;
+
+import com.hazelcast.internal.cluster.MemberInfo;
+import com.hazelcast.jet.config.JobConfig;
+import com.hazelcast.jet.core.DAG;
+
+import java.util.Set;
+
+public interface JobInvocationObserver {
```

Comment:
you made this public API. Is that intended?

Reply:
Yes, I want this interface to be available everywhere.

Reply:
if so, this requires much more thinking (eg. do we want method specific to light jobs or general) and will have to keep backward compatibility. Public API also has to have @since and full javadoc (also for methods). Currently you documented public interface by using private implementation details - this will not be understandable. You also did not document mutability of paramters and blocking/cooperativeness requirements for the listener.

We can keep it as public API but annotate it as `@Beta`. I am almost sure that someone will (mis)use this API to implement some kind of Jet job monitoring system.

Reply:
BTW, observer is public API, but method to register it is private - it does not make sense.
I changed my mind: move the observer to `impl` package. If we really need to make it public, we always can do it.